File: WarmUp/WarmUp-TD.py
#Tai Doan
#Computer Networks
#Warm up assignment
#Python

import re


def Hello():
    name = input("What's your name? ")
    print("Hello " , name , ", nice to meet you!")

def LowerLine():
    fileName = input("Please type in file name(with .txt): ")
    file = open(fileName, 'r').read()
    file = file.lower()
    file = re.sub(r"[\n\t\s]*", "", file)
    print(file)
    for i in range(97,123):
        print(chr(i),file.count(chr(i)))
    for i in range(48,58):
        print(chr(i), file.count(chr(i)))

def ReverseList(list):
    rev_list = list[::-1]
    return rev_list

class Node:
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

class LinkedList:
    def __init__(self):
            self.head = None
    def __add__(self,data):
        nNode = Node(data)
        nNode.next = self.head
        self.head = nNode
    def CycleLinkedList(self):
        key = set()
        temp = self.head
        while (temp):
            if (temp in key):
                return 1
            else:
                key.add(temp)
                temp=temp.next
        return 0

#Test function Hello()
Hello()

#Test function LowerLine()
#file name to put in: Fn2Text.txt
LowerLine()

#Test function ReverseList(list)
l = ['1','2','3','hello','my','friend'] #original list
l2 =[]                                  #reversed list
l2 = ReverseList(l)
print(l)
print(l2)

#Test function CycleLinkedList()
l = LinkedList()
l.__add__(1)
l.__add__(10)
l.__add__(2)
l.__add__(4)

l1 = LinkedList()
l1.__add__(2)
l1.__add__(4)
l1.__add__(6)

#l is a cycle linked list
l.head.next.next.next.next = l.head.next.next
if(l.CycleLinkedList() == 0):
    print("List l is not a cycle linked list")
else:
    print("List l is a cycle linked list")

#l1 is not a cycle linked list
if(l1.CycleLinkedList() == 0):
    print("List l1 is not a cycle linked list")
else:
    print("List l1 is a cycle linked list")

File: WarmUp/WarmUp-TD.cpp
//Tai Doan
//Computer Networks
//Warm up assignment
//C++

#include <iostream>
#include <string>
#include <fstream>
#include <cctype>
#include <map>

using namespace std;

void Hello() {
  string name;
  cout << "What's your name?";
  cin >> name;
  cout << endl;
  cout <<"Hello "<<name<<", nice to meet you!"<<endl;
}

void LowerLine() {
  string fileName;
  cout << "Enter file name: ";
  cin >> fileName;
  cout << endl;
  ifstream inFile(fileName);
  if (inFile.fail()) {
    cerr << "Your file did not work" << endl;
    exit(1);
  }
  string line;
  string text;
  cout << "OG line: " << line<< endl;
  cout << "OG text: " << text<<endl;
  while (inFile >> line) {
    text+=line;
  }
  cout << "Text: ";
  for (int i=0;i<text.length();i++)
    text.at(i) = tolower(text.at(i));
  cout << text << endl;
  char character[36];
  int freq[36]={0};

  for (int i =0;i<26;i++)
    character[i] = char(i+97);
  for (int j=26;j<36;j++)
    character[j] = char(j+22);
  cout << endl;

  for (int i=0;i<text.length();i++) {
    for (int j=0;j<36;j++) {
      if (text.at(i) == character[j]) {
        freq[j]++;
      }
    }
  }

  for (int i=0;i<36;i++) {
    cout << character[i] << " " << freq[i] << endl;
  }

  inFile.close();
}

int* ReverseArray(int array[],int size) {
  int* revArray = new int[size];
  for (int i = size;i>0;i--) {
    revArray[size-i] = array[i-1];
  }
  return revArray;
}

struct Node {
  int data;
  struct Node* next;
};

void insert(struct Node** head, int newData) {
  struct Node* nNode = new Node;
  nNode->data = newData;
  nNode->next = *head;
  *head = nNode;
}

int CheckCycleList(struct Node *l) {
  if (l == nullptr)
    return 1;
  struct Node *slow = l ,*fast=l;
  while (slow && fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
    if (slow == fast) {
      return 1;
    }
  }
  return 0;
}



int main() {
  //Test function Hello()
  Hello();

  //Test function LowerLine
  //file name is "Fn2Text.txt"
  LowerLine();

  //Test function ReverseArray()
  int test[5] = {1,2,3,5,6};
  int size = sizeof(test)/sizeof(test[0]);
  int* rev = ReverseArray(test,size);
  cout << "Original array: ";
  for (int i=0;i<size;i++)
    cout << test[i] << " ";
  cout << endl << "Reversed Array: ";
  for (int i=0;i<size;i++)
    cout << rev[i] << " ";
  cout << endl;

  //Test function CheckCycleList()
  struct Node*head = nullptr;
  insert(&head,1);
  insert(&head,3);
  insert(&head,5);
  insert(&head,7);

  head->next->next->next = head; //this is a cycle linked list
  int result = CheckCycleList(head);
  if(result == 0)
      cout<< "Not a cycle linked list" << endl;
  else
      cout<< "This is a cycle linked list" << endl;

  return 0;
}
